File: jarvis-desktop/tests/plain-errors.mjs
/**
 * Plain words when something goes wrong, and "How Jarvis talks" (the
 * creativity audit, 2026-09-25; JARVIS-API.md sections 4 and 24;
 * src/plain-errors.js, src/manner.js, src/manner-settings.js,
 * src-tauri/src/plain_errors.rs).
 *
 * What must hold:
 * - every word is the contract file's (tests/fixtures/plain-error-cases.json,
 *   made by tools/gen_plain_error_cases.py - the phone checks the same file):
 *   the kinds, the buttons, the waits, the codes, the manner words - and the
 *   Rust side's own copies too;
 * - each failure picks the kind the contract says (the same cases the
 *   phone's classifier runs);
 * - Details never shows a token, key, password, email address or Windows
 *   user name, and keeps what a bug report needs;
 * - the answer card: a failure shows the plain words, ONE fix button and
 *   Details; "Waking up the model…" while the PC says it is loading;
 *   "Answering…", never "Streaming", and no chunk count;
 * - Settings -> How Jarvis talks: the PC's words, ONE change per tap, no
 *   card either way, greyed on a stale link;
 * - CONTROL: the manner commands sit with the Settings window only, the fix
 *   button's command with the quickbar only, and it opens two places.
 */
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import {
  BUTTONS, classify, CODES, DETAILS_MAX, ERROR_LINE_PREFIX, fromChatFailure, fromStreamError,
  KINDS, scrubDetails, shown, STATUSES,
} from "../src/plain-errors.js";
import { DETAIL, LABEL, MANNERS, readManner, SAID, SPOKEN, TITLE, WHY } from "../src/manner.js";
import * as K from "./uikit.mjs";

const HERE = dirname(fileURLToPath(import.meta.url));
const read = (p) => readFileSync(join(HERE, "..", p), "utf8");
const CASES = JSON.parse(read("tests/fixtures/plain-error-cases.json"));

const fails = [];
const check = async (name, fn) => {
  try { await fn(); console.log(`ok    ${name}`); }
  catch (e) { fails.push(name); console.log(`FAIL  ${name}\n      ${e.message}`); }
};

/* ── The words ────────────────────────────────────────────────────────── */

await check("every kind's words are the contract's, word for word", async () => {
  assert.deepEqual(Object.keys(KINDS).sort(), Object.keys(CASES.kinds).sort());
  for (const [kind, k] of Object.entries(CASES.kinds)) {
    assert.deepEqual({ ...KINDS[kind] }, k, kind);
  }
  assert.deepEqual({ ...BUTTONS }, CASES.buttons);
  assert.deepEqual({ ...STATUSES }, CASES.statuses);
  assert.deepEqual({ ...CODES }, CASES.codes);
  assert.equal(DETAILS_MAX, CASES.details_max);
});

await check("no word a beginner would not know in what is shown", async () => {
  for (const [kind, k] of Object.entries(CASES.kinds)) {
    const text = `${k.says} ${k.fix} ${k.button}`;
    assert.ok(!/\b(HTTP|503|404|stale|chunk|stream|exception|errno|socket|tier)\b/i.test(text),
      `${kind}: ${text}`);
    if (kind !== "pc_said") {
      assert.ok(k.says && k.fix, `${kind} has what happened AND what to do`);
    }
  }
  assert.ok(!/stream/i.test(Object.values(STATUSES).join(" ")));
});

await check("the Rust side's own copies are the contract's words", async () => {
  const rs = read("src-tauri/src/plain_errors.rs");
  const konst = (name) => {
    const m = rs.match(new RegExp(`const ${name}: &str =\\s*"((?:[^"\\\\]|\\\\.)*)"`, "s"));
    assert.ok(m, name);
    return m[1].replace(/\\\n\s*/g, "").replace(/\\"/g, "\"");
  };
  const pairs = { NOT_RUNNING: "jarvis_not_running", TIMEOUT: "timeout", DROPPED: "connection_dropped",
                  KEY_STORE: "key_store_refused" };
  for (const [name, kind] of Object.entries(pairs)) {
    assert.equal(konst(`${name}_SAYS`), CASES.kinds[kind].says, name);
    assert.equal(konst(`${name}_FIX`), CASES.kinds[kind].fix, name);
  }
  assert.ok(rs.includes(`"${ERROR_LINE_PREFIX.replace("\u001f", "\\u{1f}")}`));
});

await check("each failure picks the kind the contract says (the phone runs the same cases)", async () => {
  for (const c of CASES.classify) {
    assert.equal(classify(c.input), c.kind, c.name);
  }
});

await check("the PC's own sentence is shown as it is, first letter raised", async () => {
  const s = shown("pc_said", "the second card is not detected");
  assert.equal(s.says, "The second card is not detected.");
  assert.equal(s.button, "");
  assert.equal(shown("model_missing").button, "Choose a model");
});

/* ── Details ──────────────────────────────────────────────────────────── */

await check("Details: every secret gone, what a bug report needs kept", async () => {
  for (const c of CASES.scrub) {
    const out = scrubDetails(c.input);
    for (const g of c.gone) assert.ok(!out.includes(g), `${g} survived: ${out}`);
    for (const k of c.kept) assert.ok(out.includes(k), `${k} was lost: ${out}`);
  }
  const t = CASES.scrub_known_token;
  const out = scrubDetails(t.input, t.token);
  for (const g of t.gone) assert.ok(!out.includes(g), out);
  for (const k of t.kept) assert.ok(out.includes(k), out);
  assert.ok(scrubDetails("x".repeat(5000)).length <= CASES.details_max);
});

await check("a failed chat's tagged facts become the plain words, the detail scrubbed", async () => {
  const secret = "tv" + "ly-" + "abcdefghijklmnop";
  const line = ERROR_LINE_PREFIX + JSON.stringify({ http: 503, detail: `key ${secret}` });
  const p = fromChatFailure(line);
  assert.equal(p.kind, "feature_off");
  assert.equal(p.says, CASES.kinds.feature_off.says);
  assert.ok(p.details.includes("HTTP 503") && !p.details.includes(secret), p.details);
  const q = fromChatFailure(ERROR_LINE_PREFIX + JSON.stringify({ network: "refused", detail: "x" }));
  assert.equal(q.kind, "jarvis_not_running");
  // An older build's plain sentence is shown as it is.
  assert.equal(fromChatFailure("Temporary chat isn't available.").kind, "pc_said");
  const s = fromStreamError({ message: "The local model is not running.", code: "model_not_running" });
  assert.equal(s.says, CASES.kinds.model_not_running.says);
  assert.ok(s.details.includes("The local model is not running."));
});

/* ── The manner words ─────────────────────────────────────────────────── */

await check("How Jarvis talks: the PC's words, word for word", async () => {
  const m = CASES.manner;
  assert.equal(TITLE, m.title);
  assert.equal(DETAIL, m.detail);
  assert.equal(SPOKEN, m.spoken);
  assert.deepEqual([...MANNERS], m.choices.map((c) => c.id));
  for (const c of m.choices) {
    assert.equal(LABEL[c.id], c.label, c.id);
    assert.equal(WHY[c.id], c.why, c.id);
  }
  assert.deepEqual({ ...SAID }, m.said);
  assert.equal(m.default, "warm");
  const v = readManner({ available: true, manner: "plain", choices: m.choices });
  assert.equal(v.manner, "plain");
  assert.equal(readManner({ choices: [] }).manner, "warm", "no setting in the answer: the default");
  assert.equal(readManner(null).available, false, "no answer: never drawn as a choice");
  assert.equal(readManner({ available: false, why: "old" }).available, false);
});

/* ── The answer card ──────────────────────────────────────────────────── */

const { base, close } = await K.serve();
const browser = await K.launch();

async function ask(replies, before) {
  const page = await K.open(browser, base, "index.html", { chatReplies: [replies] },
    { width: 750, height: 600 });
  await page.evaluate(() => {
    window.__statuses = [];
    const el = document.getElementById("card-status-text");
    new MutationObserver(() => window.__statuses.push(el.textContent))
      .observe(el, { childList: true, characterData: true, subtree: true });
  });
  if (before) await before(page);
  await page.locator("#prompt").fill("hi");
  await page.locator("#prompt").press("Enter");
  await page.waitForTimeout(400);
  const got = await page.evaluate(() => ({
    answer: document.getElementById("answer").textContent.trim(),
    status: document.getElementById("card-status-text").textContent,
    stat: document.getElementById("card-stat").textContent,
    statuses: window.__statuses,
    problemShown: !document.getElementById("answer-problem").hidden,
    button: document.getElementById("problem-action").hidden
      ? "" : document.getElementById("problem-action").textContent,
    details: document.getElementById("problem-details-text").textContent,
  }));
  return { page, got };
}

await check("an answer says Answering…, never Streaming, and counts no chunks", async () => {
  const { page, got } = await ask(["data: {\"choices\":[{\"delta\":{\"content\":\"Hello.\"}}]}",
    "data: [DONE]"]);
  await page.close();
  assert.ok(got.statuses.includes(STATUSES.answering), JSON.stringify(got.statuses));
  assert.ok(!got.statuses.some((s) => /Streaming/.test(s)));
  assert.equal(got.stat, "");
  assert.equal(got.problemShown, false);
});

await check("the PC says the model is loading: Waking up the model…", async () => {
  const { page, got } = await ask([": jarvis-status loading",
    "data: {\"choices\":[{\"delta\":{\"content\":\"Morning.\"}}]}", "data: [DONE]"]);
  await page.close();
  assert.ok(got.statuses.includes(STATUSES.loading), JSON.stringify(got.statuses));
});

await check("a failure: the plain words, ONE fix button, the PC's sentence behind Details", async () => {
  const body = "data: " + JSON.stringify({ error: { message: "The model “qwen3:14b” is not "
    + "installed on this PC.", type: "jarvis", code: "model_missing" } });
  const { page, got } = await ask([body], async (pg) => {
    await pg.evaluate(() => {
      window.__opened = [];
      const core = window.__TAURI__.core;
      const inner = core.invoke;
      core.invoke = async (cmd, args) => {
        if (cmd === "open_fix_place") { window.__opened.push(args.place); return null; }
        return inner(cmd, args);
      };
    });
  });
  assert.ok(got.answer.includes(CASES.kinds.model_missing.says), got.answer);
  assert.ok(got.answer.includes(CASES.kinds.model_missing.fix), got.answer);
  assert.equal(got.button, "Choose a model");
  assert.ok(got.details.includes("qwen3:14b"), got.details);
  // Selectable for a bug report (ease-of-use audit #6); the rest of the bar is not.
  const select = await page.evaluate(() => [
    getComputedStyle(document.getElementById("problem-details-text")).userSelect,
    getComputedStyle(document.body).userSelect]);
  assert.deepEqual(select, ["text", "none"]);
  await page.locator("#problem-action").click();
  const opened = await page.evaluate(() => window.__opened);
  await page.close();
  assert.deepEqual(opened, ["brain"]);
});

await check("Try again sends the same question, with the same tag", async () => {
  const line = ERROR_LINE_PREFIX + JSON.stringify({ network: "refused", detail: "x" });
  const page = await K.open(browser, base, "index.html",
    // The first stream_chat is refused below before it reads a reply, so
    // the one scripted reply is the second call's.
    { chatReplies: [["data: {\"choices\":[{\"delta\":{\"content\":\"Back.\"}}]}", "data: [DONE]"]] },
    { width: 750, height: 600 });
  await page.evaluate((l) => {
    window.__asked = [];
    const core = window.__TAURI__.core;
    const inner = core.invoke;
    let first = true;
    core.invoke = async (cmd, args) => {
      if (cmd === "stream_chat") {
        window.__asked.push(JSON.stringify(args.messages[args.messages.length - 1]));
        if (first) { first = false; throw new Error(l); }
      }
      return inner(cmd, args);
    };
  }, line);
  await page.locator("#prompt").fill("what's on today");
  await page.locator("#prompt").press("Enter");
  await page.waitForTimeout(300);
  const said = await page.locator("#answer").textContent();
  assert.ok(said.includes(CASES.kinds.jarvis_not_running.says), said);
  assert.equal(await page.locator("#problem-action").textContent(), "Try again");
  await page.locator("#problem-action").click();
  await page.waitForTimeout(300);
  const asked = await page.evaluate(() => window.__asked);
  const after = await page.locator("#answer").textContent();
  const problemHidden = await page.locator("#answer-problem").isHidden();
  await page.close();
  assert.equal(asked.length, 2);
  assert.equal(asked[0], asked[1], "the same words and the same provenance");
  assert.ok(after.includes("Back."), after);
  assert.ok(problemHidden, "the old fix button stays after a good answer");
});

await check("\"Jarvis isn't running\": a second button opens Settings at Starting Jarvis for you", async () => {
  // Ease-of-use audit 2026-09-27 #2: the fix names the real place, and the
  // desktop opens it - "More options" open, the card scrolled to.
  const line = ERROR_LINE_PREFIX + JSON.stringify({ network: "refused", detail: "x" });
  const page = await K.open(browser, base, "index.html", {}, { width: 750, height: 600 });
  await page.evaluate((l) => {
    window.__opened = [];
    const core = window.__TAURI__.core;
    const inner = core.invoke;
    core.invoke = async (cmd, args) => {
      if (cmd === "stream_chat") throw new Error(l);
      if (cmd === "open_fix_place") { window.__opened.push(args.place); return null; }
      return inner(cmd, args);
    };
  }, line);
  await page.locator("#prompt").fill("hi");
  await page.locator("#prompt").press("Enter");
  await page.waitForTimeout(300);
  const said = await page.locator("#answer").textContent();
  assert.ok(said.includes("More options") && said.includes("Starting Jarvis for you"), said);
  assert.equal(await page.locator("#problem-action").textContent(), "Try again");
  assert.equal(await page.locator("#problem-where").isVisible(), true);
  await page.locator("#problem-where").click();
  const opened = await page.evaluate(() => window.__opened);
  const left = await page.evaluate(() => JSON.parse(localStorage.getItem("jarvis.settings.place")));
  await page.close();
  assert.deepEqual(opened, ["settings"]);
  assert.equal(left.place, "start-jarvis");

  const sp = await K.open(browser, base, "settings.html", {}, { width: 900, height: 900 });
  await sp.evaluate(() => localStorage.setItem("jarvis.settings.place",
    JSON.stringify({ place: "start-jarvis", at: Date.now() })));
  await sp.reload();
  await sp.waitForTimeout(400);
  const got = await sp.evaluate(() => ({
    open: document.getElementById("more-options").open,
    top: document.getElementById("start-jarvis").getBoundingClientRect().top,
    kept: localStorage.getItem("jarvis.settings.place"),
  }));
  await sp.close();
  assert.equal(got.open, true, "More options opens");
  assert.ok(got.top >= -2 && got.top < 200, `the card is in view (top ${got.top})`);
  assert.equal(got.kept, null, "the place is taken once");
});

await check("another failure shows no \"Show me where\"", async () => {
  const body = "data: " + JSON.stringify({ error: { message: "x", type: "jarvis", code: "model_missing" } });
  const { page } = await ask([body]);
  const where = await page.locator("#problem-where").isVisible();
  await page.close();
  assert.equal(where, false);
});

/* ── Settings -> How Jarvis talks ─────────────────────────────────────── */

/** The bridge for the two manner commands, on top of uikit's. */
function mannerBridge({ m, choices }) {
  const core = window.__TAURI__.core;
  const inner = core.invoke;
  window.__manner = { manner: m, sets: [] };
  core.invoke = async (cmd, args) => {
    if (cmd === "get_manner") {
      return { available: true, manner: window.__manner.manner, default: "warm",
               title: "How Jarvis talks", detail: "PC detail", spoken: "PC spoken", choices };
    }
    if (cmd === "set_manner") {
      window.__manner.sets.push(args);
      window.__manner.manner = args.manner;
      return { ok: true, said: "PC said it." };
    }
    return inner(cmd, args);
  };
}

async function settings(manner = "warm", data = {}) {
  const page = await K.open(browser, base, "settings.html", data, { width: 900, height: 1800 });
  await page.addInitScript(mannerBridge, { m: manner, choices: CASES.manner.choices });
  await page.reload();
  await page.waitForTimeout(700);
  return page;
}

await check("Settings shows the PC's words for both choices, with the one in use ticked", async () => {
  const page = await settings("warm");
  const got = await page.evaluate(() => ({
    rows: [...document.querySelectorAll("#mn-choices .theme-row")].map((r) => ({
      id: r.dataset.choice, text: r.textContent, checked: r.querySelector("input").checked })),
    detail: document.getElementById("mn-detail").textContent,
    spoken: document.getElementById("mn-spoken").textContent,
    heading: document.querySelector("#manner h2").textContent,
  }));
  await page.close();
  assert.equal(got.heading, CASES.manner.title);
  assert.equal(got.detail, "PC detail");
  assert.equal(got.spoken, "PC spoken");
  assert.deepEqual(got.rows.map((r) => r.id), ["warm", "plain"]);
  assert.ok(got.rows[0].checked && !got.rows[1].checked);
  for (const [i, c] of CASES.manner.choices.entries()) {
    assert.ok(got.rows[i].text.includes(c.label) && got.rows[i].text.includes(c.why), c.id);
  }
});

await check("one tap sends ONE change, and no card is mentioned", async () => {
  const page = await settings("warm");
  await page.locator('#mn-choices input[value="plain"]').check();
  await page.waitForTimeout(300);
  const got = await page.evaluate(() => ({
    sets: window.__manner.sets,
    status: document.getElementById("mn-status").textContent,
    checked: document.querySelector('#mn-choices input[value="plain"]').checked,
  }));
  await page.close();
  assert.deepEqual(got.sets, [{ manner: "plain" }]);
  assert.equal(got.status, "PC said it.");
  assert.ok(got.checked);
  assert.ok(!/approv|card/i.test(read("src/manner-settings.js").split("*/").slice(1).join("")),
    "the page never talks about a card");
});

await check("greyed on a stale link", async () => {
  const page = await settings("warm", { link: { stale: true } });
  const disabled = await page.evaluate(() =>
    [...document.querySelectorAll("#mn-choices input")].every((i) => i.disabled));
  await page.close();
  assert.ok(disabled);
});

/* ── Control ──────────────────────────────────────────────────────────── */

await check("the manner commands are Settings-only; the fix button's is the quickbar's", async () => {
  const sets = read("src-tauri/permissions/surfaces.toml").split("[[set]]").slice(1);
  const holders = (cmd) => sets.filter((s) => s.includes(`"allow-${cmd.replace(/_/g, "-")}"`))
    .map((s) => s.match(/identifier = "([^"]+)"/)[1]);
  assert.deepEqual(holders("get_manner"), ["settings-surface"]);
  assert.deepEqual(holders("set_manner"), ["settings-surface"]);
  assert.deepEqual(holders("open_fix_place"), ["quickbar-surface"]);
  const rs = read("src-tauri/src/plain_errors.rs");
  const fix = rs.slice(rs.indexOf("pub fn open_fix_place"));
  assert.match(fix, /"settings" =>/);
  assert.match(fix, /"brain" =>/);
  assert.match(fix, /_ => Err/);
  const set = rs.slice(rs.indexOf("pub async fn set_manner"), rs.indexOf("pub fn open_fix_place"));
  assert.match(set, /link\(\)\.stale/, "set_manner is held on a stale link");
});

await browser.close();
await close();
if (fails.length) {
  console.log(`\n${fails.length} failed`);
  process.exit(1);
}
console.log("\nall passed");

File: jarvis-desktop/src-tauri/src/plain_errors.rs
//! Plain words when something goes wrong, and "How Jarvis talks" (the
//! creativity audit, 2026-09-25: item 5, and the owner's manner decision).
//!
//! The words themselves live in ONE list, `tools/gen_plain_error_cases.py`,
//! written to `tests/fixtures/plain-error-cases.json` (and a byte-identical
//! copy the phone reads). The page's `src/plain-errors.js` holds all of them;
//! the Rust side says only the three a request that never reached Jarvis can
//! end in ([`unreachable_words`]), and tags a failed chat so the page can
//! pick the words itself ([`chat_failure`]). The tests below fail if a word
//! here differs from the file.
//!
//! Two commands for Settings -> "How Jarvis talks" (backend/manner.patch,
//! jarvis_manner.py; JARVIS-API.md section 27), settings window only:
//!
//! * [`get_manner`] - `GET /api/manner`: "warm" (the default) or "plain",
//!   with the PC's own words for both.
//! * [`set_manner`] - `POST /api/manner {"manner": ...}`: at once, NO
//!   approval card either way - it changes only how answers are worded,
//!   never what Jarvis does, asks, remembers or sends. Held on a stale link
//!   all the same (rule 4: nothing is sent while the link is stale).
//!
//! And one for the quickbar: [`open_fix_place`] opens Settings or the Brain
//! when the owner presses an error's fix button ("Check the connection
//! settings", "Choose a model"). It opens a window and nothing else; the
//! app lock still applies (windows.rs).

use std::time::Duration;

use tauri::{AppHandle, Manager};

use crate::commands::{
    backend_refusal, backend_unreachable, jarvis_base, jarvis_client, jarvis_headers,
};

/// `jarvis_not_running`: nothing listens at the address (connection refused).
pub(crate) const NOT_RUNNING_SAYS: &str = "Jarvis isn't running on your PC.";
pub(crate) const NOT_RUNNING_FIX: &str = "The PC is on, but Jarvis is not started. On the PC, \
     open Jarvis Desktop's Settings, then More options, and press Start under \"Starting Jarvis \
     for you\" (it needs \"Let Jarvis Desktop start and stop Jarvis\" on). Or start it in \
     PowerShell, the way you set it up. Then try again.";
/// `timeout`: connected, then no answer in time.
pub(crate) const TIMEOUT_SAYS: &str = "Jarvis took too long to answer.";
pub(crate) const TIMEOUT_FIX: &str =
    "Try again in a moment. If it keeps happening, restart Ollama on the PC.";
/// `connection_dropped`: the request broke on the way.
pub(crate) const DROPPED_SAYS: &str = "The connection to your PC dropped.";
pub(crate) const DROPPED_FIX: &str =
    "Try again. If it keeps happening, check the private network is steady at both ends.";

/// `key_store_refused`: Credential Manager would not save a key.
pub(crate) const KEY_STORE_SAYS: &str = "Windows Credential Manager wouldn't save it.";
pub(crate) const KEY_STORE_FIX: &str = "Nothing was written anywhere else. Try again; if it \
     keeps failing, restart the PC and try once more.";

/// The sentence for a key Credential Manager refused, with its reason (a
/// code from Windows, never the key) for a bug report.
pub(crate) fn key_store_words(reason: &str) -> String {
    format!("{KEY_STORE_SAYS} {KEY_STORE_FIX} (Details: {reason}.)")
}

/// Starts a failed chat's one line, instead of a sentence: what the page
/// needs to pick the plain words (plain-errors.js `fromChatFailure`). A
/// unit-separator control character never starts a real sentence.
pub const ERROR_LINE_PREFIX: &str = "\u{1f}jarvis-error:";

/// The network failure's kind, in the contract file's words.
pub(crate) fn network_kind(is_connect: bool, is_timeout: bool) -> &'static str {
    if is_connect && is_timeout {
        "connect_timeout"
    } else if is_connect {
        "refused"
    } else if is_timeout {
        "read_timeout"
    } else {
        "dropped"
    }
}

/// The sentence for a request that never got an answer from Jarvis.
pub(crate) fn unreachable_words(is_connect: bool, is_timeout: bool) -> String {
    let (says, fix) = match network_kind(is_connect, is_timeout) {
        "refused" | "connect_timeout" => (NOT_RUNNING_SAYS, NOT_RUNNING_FIX),
        "read_timeout" => (TIMEOUT_SAYS, TIMEOUT_FIX),
        _ => (DROPPED_SAYS, DROPPED_FIX),
    };
    format!("{says} {fix}")
}

/// A failed chat, as the one tagged line the page reads: the network kind
/// or the HTTP status, the PC's own sentence when it sent one, and the
/// technical detail for the page's "Details" (which it scrubs).
pub(crate) fn chat_failure(
    network: Option<&str>,
    http: Option<u16>,
    said: Option<&str>,
    detail: &str,
) -> String {
    let mut out = serde_json::Map::new();
    if let Some(n) = network {
        out.insert("network".into(), serde_json::json!(n));
    }
    if let Some(code) = http {
        out.insert("http".into(), serde_json::json!(code));
    }
    if let Some(s) = said.map(str::trim).filter(|s| !s.is_empty()) {
        out.insert("said".into(), serde_json::json!(s));
    }
    // Bounded: a whole HTML error page is not a detail.
    let detail: String = detail.chars().take(1000).collect();
    out.insert("detail".into(), serde_json::json!(detail));
    format!("{ERROR_LINE_PREFIX}{}", serde_json::Value::Object(out))
}

const MANNER_PATH: &str = "/api/manner";
const MANNERS: [&str; 2] = ["warm", "plain"];
const READ_TIMEOUT: Duration = Duration::from_secs(15);
const WRITE_TIMEOUT: Duration = Duration::from_secs(20);

/// What a backend without jarvis_manner.py is told. The phone says the same.
pub(crate) const MANNER_MISSING: &str =
    "Your PC's Jarvis does not have this setting yet - run apply-patches.ps1 on the PC.";

const STALE: &str =
    "The connection to Jarvis is catching up, so nothing can be sent until it does.";

const UNREADABLE: &str = "Jarvis answered, but not in a way this app can read. Update the \
                          backend by running apply-patches.ps1.";

fn missing(code: u16, body: &str) -> bool {
    code == 404
        || code == 501
        || (code == 503
            && serde_json::from_str::<serde_json::Value>(body)
                .ok()
                .and_then(|v| v.get("available").and_then(|a| a.as_bool()))
                == Some(false))
}

/// [`get_manner`]'s reading of the answer, tested against the contract file.
pub(crate) fn manner_answer(status: u16, body: &str) -> Result<serde_json::Value, String> {
    if (200..300).contains(&status) {
        return serde_json::from_str::<serde_json::Value>(body)
            .ok()
            .filter(|v| v.get("choices").is_some_and(|c| c.is_array()))
            .ok_or_else(|| UNREADABLE.to_string());
    }
    if missing(status, body) {
        return Ok(serde_json::json!({ "available": false, "why": MANNER_MISSING }));
    }
    Err(backend_refusal(status, body))
}

/// The body of the one change: "warm" or "plain", nothing else.
pub(crate) fn manner_body(manner: &str) -> Result<serde_json::Value, String> {
    if !MANNERS.contains(&manner) {
        return Err("That is not one of the two choices.".to_string());
    }
    Ok(serde_json::json!({ "manner": manner }))
}

/// "How Jarvis talks": `GET /api/manner`.
#[tauri::command]
pub async fn get_manner(app: AppHandle) -> Result<serde_json::Value, String> {
    let base = jarvis_base(&app);
    let response = jarvis_client(Some(READ_TIMEOUT))?
        .get(format!("{base}{MANNER_PATH}"))
        .headers(jarvis_headers(&app)?)
        .send()
        .await
        .map_err(|e| backend_unreachable(&e, &base))?;
    let status = response.status().as_u16();
    let body = response.text().await.unwrap_or_default();
    manner_answer(status, &body)
}

/// Warm or plain, at once, with no card. Held on a stale link.
#[tauri::command]
pub async fn set_manner(app: AppHandle, manner: String) -> Result<serde_json::Value, String> {
    let body = manner_body(&manner)?;
    if app.state::<crate::stream::StreamState>().link().stale {
        return Err(STALE.to_string());
    }
    let base = jarvis_base(&app);
    let response = jarvis_client(Some(WRITE_TIMEOUT))?
        .post(format!("{base}{MANNER_PATH}"))
        .headers(jarvis_headers(&app)?)
        .json(&body)
        .send()
        .await
        .map_err(|e| backend_unreachable(&e, &base))?;
    let status = response.status().as_u16();
    let text = response.text().await.unwrap_or_default();
    if (200..300).contains(&status) {
        return serde_json::from_str::<serde_json::Value>(&text)
            .ok()
            .filter(|v| v.is_object())
            .ok_or_else(|| UNREADABLE.to_string());
    }
    if missing(status, &text) {
        return Err(MANNER_MISSING.to_string());
    }
    Err(backend_refusal(status, &text))
}

/// The place an error's fix button opens: "settings" or "brain". Nothing
/// else is accepted.
#[tauri::command]
pub fn open_fix_place(app: AppHandle, place: String) -> Result<(), String> {
    match place.as_str() {
        "settings" => crate::windows::show_settings(&app),
        "brain" => crate::windows::show_brain(&app),
        _ => Err("That is not a place this button opens.".to_string()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The one list of words, made by `tools/gen_plain_error_cases.py`.
    const CASES: &str = include_str!("../../tests/fixtures/plain-error-cases.json");

    fn cases() -> serde_json::Value {
        serde_json::from_str(CASES).expect("plain-error-cases.json is JSON")
    }

    fn words(kind: &str) -> (String, String) {
        let doc = cases();
        let k = &doc["kinds"][kind];
        (
            k["says"].as_str().expect("says").to_string(),
            k["fix"].as_str().expect("fix").to_string(),
        )
    }

    #[test]
    fn the_rust_words_are_the_contract_words() {
        assert_eq!(
            words("jarvis_not_running"),
            (NOT_RUNNING_SAYS.into(), NOT_RUNNING_FIX.into())
        );
        assert_eq!(words("timeout"), (TIMEOUT_SAYS.into(), TIMEOUT_FIX.into()));
        assert_eq!(
            words("connection_dropped"),
            (DROPPED_SAYS.into(), DROPPED_FIX.into())
        );
        assert_eq!(
            words("key_store_refused"),
            (KEY_STORE_SAYS.into(), KEY_STORE_FIX.into())
        );
    }

    #[test]
    fn a_network_failure_is_classified_as_the_contract_says() {
        let doc = cases();
        for case in doc["classify"].as_array().expect("classify") {
            let Some(net) = case["input"]["network"].as_str() else {
                continue;
            };
            let flags = match net {
                "refused" => (true, false),
                "connect_timeout" => (true, true),
                "read_timeout" => (false, true),
                "dropped" => (false, false),
                _ => continue, // names and routes: the phone's
            };
            assert_eq!(network_kind(flags.0, flags.1), net);
        }
    }

    #[test]
    fn the_chat_failure_line_carries_facts_not_words() {
        let line = chat_failure(Some("refused"), None, None, "error sending request");
        let json = line.strip_prefix(ERROR_LINE_PREFIX).expect("tagged");
        let v: serde_json::Value = serde_json::from_str(json).expect("json");
        assert_eq!(v["network"], "refused");
        assert!(v.get("said").is_none());
        let line = chat_failure(None, Some(503), Some("  the model is loading "), "x");
        let v: serde_json::Value =
            serde_json::from_str(line.strip_prefix(ERROR_LINE_PREFIX).unwrap()).unwrap();
        assert_eq!(v["http"], 503);
        assert_eq!(v["said"], "the model is loading");
    }

    #[test]
    fn manner_choices_and_bodies() {
        assert!(manner_body("warm").is_ok());
        assert!(manner_body("plain").is_ok());
        assert!(manner_body("rude").is_err());
        let doc = cases();
        let ids: Vec<&str> = doc["manner"]["choices"]
            .as_array()
            .expect("choices")
            .iter()
            .map(|c| c["id"].as_str().unwrap())
            .collect();
        assert_eq!(ids, MANNERS);
        assert!(manner_answer(404, "").unwrap()["available"] == false);
        assert!(manner_answer(200, "{}").is_err());
    }
}
